src/util/bundleUtils.ts

```diff
@@ -14,3 +20,90 @@ export function createSearchsetBundle<T extends fhir4.FhirResource>(entries: T[]
     entry: entries.map(e => ({ resource: e }))
   };
 }
+
+/**
+ * Takes in a measure resource, finds all dependent library resources and bundles them
+ * together with the measure in a collection bundle
+ */
+export async function createMeasurePackageBundle(measure: fhir4.Measure): Promise<fhir4.Bundle<fhir4.FhirResource>> {
+  logger.info(`Assembling collection bundle from Measure ${measure.id}`);
+  if (measure.library && measure.library.length > 0) {
+    const [mainLibraryRef] = measure.library;
+    const mainLibQuery = getQueryFromReference(mainLibraryRef);
+    const libs = await findResourcesWithQuery(mainLibQuery, 'Library');
+    if (!libs || libs.length < 1) {
+      throw new ResourceNotFoundError(`Could not find Library ${mainLibraryRef} referenced by Measure ${measure.id}`);
+    }
+    const mainLib = libs[0];
+
+    const allLibsDups = await getAllDependentLibraries(mainLib as fhir4.Library);
+    // de-dup by id using map
+    const idMap = new Map(allLibsDups.map(lib => [lib.id, lib]));
+    const allLibs = Array.from(idMap.values());
+    const result: fhir4.Bundle = { resourceType: 'Bundle', type: 'collection', id: v4() };
+    result.entry = [{ resource: measure }];
+    result.entry.push(
+      ...allLibs.map(r => ({
+        resource: r
+      }))
+    );
+    return result;
+  } else {
+    throw new BadRequestError(`Uploaded measure: ${measure.id} does not reference a Library`);
+  }
+}
+
+/**
+ * Assemble a mongo query based on a reference to another resource
+ * @param {string} reference assumed to be canonical
+ * @returns {Filter} mongo query to pass in to mongo controller to search for the referenced resource
+ */
+export function getQueryFromReference(reference: string): Filter<any> {
+  if (reference.includes('|')) {
+    const [urlPart, versionPart] = reference.split('|');
+    return { url: urlPart, version: versionPart };
+  } else {
+    return { url: reference };
+  }
+}
+
+/**
+ * Iterate through relatedArtifact of library and return list of all dependent libraries used
+ */
+export async function getAllDependentLibraries(lib: fhir4.Library): Promise<fhir4.Library[]> {
+  logger.debug(`Retrieving all dependent libraries for library: ${lib.id}`);
+  const results = [lib];
+
+  // If the library has no dependencies, we are done
+  if (!lib.relatedArtifact || (Array.isArray(lib.relatedArtifact) && lib.relatedArtifact.length === 0)) {
+    return results;
+  }
+  // This filter checks for the 'Library' keyword on all related artifacts
+  const depLibUrls = lib.relatedArtifact
+    .filter(
+      ra =>
+        ra.type === 'depends-on' &&
+        ra.resource?.includes('Library') &&
+        ra.resource !== 'http://fhir.org/guides/cqf/common/Library/FHIR-ModelInfo|4.0.1'
+    ) // exclude modelinfo dependency
+    .map(ra => ra.resource as string); //TODO: may be able to improve this typing
```

Comment:
Not a reviewer but just saying that the casting to string is fine here. You know more than typescript in the sense that anything where `ra.resource` is undefined will have already been filtered out of the array so it's safe to always do `as string`. TypeScript isn't smart enough to recognize the  chaining of the array calls and how they influence each other in that regard. Can probably remove the TODO

Reply:
Thanks! Removed TODO